Jellyfin.Plugin.Tvdb/ScheduledTasks/UpdateTask.cs

```diff
@@ -0,0 +1,132 @@
+        public string Category => "TheTVDB";
+
+        private static int MetadataUpdateInHours => TvdbPlugin.Instance?.Configuration.MetadataUpdateInHours * -1 ?? -1;
+
+        /// <inheritdoc/>
+        public async Task ExecuteAsync(IProgress<double> progress, CancellationToken cancellationToken)
+        {
+            progress.Report(0);
+            _logger.LogInformation("Checking for metadata updates.");
+            var toUpdateItems = await GetItemsUpdated(cancellationToken).ConfigureAwait(false);
+            _logger.LogInformation("Found {0} items to update.", toUpdateItems.Count);
+            progress.Report(10);
+            MetadataRefreshOptions refreshOptions = new MetadataRefreshOptions(new DirectoryService(_fileSystem))
+            {
+                MetadataRefreshMode = MetadataRefreshMode.FullRefresh
+            };
+            double increment = 90.0 / toUpdateItems.Count;
+            double currentProgress = 10;
+            foreach (BaseItem item in toUpdateItems)
+            {
+                _logger.LogInformation("Refreshing metadata for TvdbId {Tvdbid}:{Name}", item.GetTvdbId(), item.Name);
+                await _providerManager.RefreshSingleItem(
+                    item,
+                    refreshOptions,
+                    cancellationToken).ConfigureAwait(false);
+                currentProgress += increment;
+                progress.Report(currentProgress);
+            }
+
+            progress.Report(100);
+        }
+
+        /// <inheritdoc/>
+        public IEnumerable<TaskTriggerInfo> GetDefaultTriggers()
+        {
+            return Enumerable.Empty<TaskTriggerInfo>();
+        }
+
+        /// <summary>
+        /// Gets all items that have been updated.
+        /// </summary>
+        /// <returns>List of items that have been updated.</returns>
+        private async Task<List<BaseItem>> GetItemsUpdated(CancellationToken cancellationToken)
+        {
+            double fromTime = DateTimeOffset.UtcNow.AddHours(MetadataUpdateInHours).ToUnixTimeSeconds();
+            IReadOnlyList<EntityUpdate> episodeUpdates = await _tvdbClientManager.GetUpdates(fromTime, cancellationToken, Type.Episodes, Action.Update).ConfigureAwait(false);
+            IReadOnlyList<EntityUpdate> seriesUpdates = await _tvdbClientManager.GetUpdates(fromTime, cancellationToken, Type.Series, Action.Update).ConfigureAwait(false);
+
+            var allUpdates = episodeUpdates.Concat(seriesUpdates);
+
+            string providerId = MetadataProvider.Tvdb.ToString();
+
+            List<BaseItem> toUpdateItems = new List<BaseItem>();
+
+            Dictionary<string, string> providerIdPair = new Dictionary<string, string>() { { providerId, string.Empty } };
+            InternalItemsQuery query = new InternalItemsQuery();
+
+            foreach (EntityUpdate update in allUpdates)
+            {
+                providerIdPair[providerId] = update.RecordId!.Value.ToString(CultureInfo.InvariantCulture);
```

Comment:
Why not populate all of the provider ids at once instead of adding 1, then querying, then adding another one?

Reply:
HasAnyProviderId expects a dictionary where the key is the MetadataProvider and the value is the TvdbId. In this loop I update the TvdbId before querying. I don't think its possible for me to a request for all TvdbId at once using the librarymanager.

Reply:
you're right.. got stuck because the repo just uses a list at the end result